src/AngleSharp.Diffing/Core/HtmlDifferenceEngine.cs

```diff
@@ -85,9 +85,14 @@ private IEnumerable<IDiff> CompareNode(in Comparison comparison)
         }
 
         var compareRes = _diffingStrategy.Compare(comparison);
-        if (compareRes.HasFlag(CompareResult.Different))
+        if (compareRes.Decision.HasFlag(CompareDecision.Different))
         {
-            IDiff diff = new NodeDiff(comparison);
+            if (compareRes.Diff != null && compareRes.Diff is not NodeDiff)
+            {
+                throw new InvalidOperationException("Comparison must return NodeDiff.");
+            }
+
+            IDiff diff = compareRes.Diff ?? new NodeDiff(comparison);
```

Comment:
ps. if I delete the `if (compareRes.Diff != null && compareRes.Diff is not NodeDiff)` check, all tests still pass.

Reply:
Because we don't want that a custom comparer returns an AttrDiff for a node comparison. It would have been possible to make the delegate more strict, but then a lot of new types would have been introduced.

Reply:
Your decision...

Reply:
If we cannot come up with a good reason, then I suggest leaving it open for now.